cishouseholds/impute.py

```diff
@@ -10,6 +10,63 @@
 from pyspark.sql.types import DoubleType
 from pyspark.sql.window import Window
 
+from cishouseholds.edit import update_column_values_from_map
+
+
+def impute_think_had_covid(
+    df: DataFrame,
+    participant_id_column: str,
+    date_column: str,
+    visit_date_column: str,
+    type_column: str,
+    contact_column: str,
+    hierarchy_map: dict,
+):
+    """
+    impute think had covid and associated variables
+    Parameters
+    ----------
+    df
+    participant_id_column
+    date_column
+    visit_date_column
+    type_column
+    contact_column
+    hierarchy_map
+    """
+    date_exists_df = df.select(participant_id_column, date_column)
+    transformed_df = df.join(
+        date_exists_df.withColumnRenamed(date_column, f"{date_column}_ref"), on=participant_id_column, how="left"
+    )
+    transformed_df = transformed_df.withColumn(
+        f"{date_column}_ref",
+        F.when(F.col(f"{date_column}_ref") <= F.col(visit_date_column), F.col(f"{date_column}_ref")).otherwise(None),
+    )
+
+    window = Window.partitionBy(participant_id_column, visit_date_column, f"{date_column}_ref")
+    transformed_df = update_column_values_from_map(transformed_df, type_column, hierarchy_map)
+    transformed_df = transformed_df.withColumn(type_column, F.min(type_column).over(window))
+
+    reverse_map = {}
```

Comment:
It's more pythonic and concise to reverse a dict using `{value: key for key, value in hierarchy_map.items()}`

Reply:
modified